### PR TITLE
Create ApplyingChunk.md

### PR DESCRIPTION
**Pull request by EgorKulikov**

### CODE REVIEW

File: specs/RuntimeSpec/ApplyingChunk.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+* If this is first block of an epoch, dispense epoch rewards to validators (in order of *validator_accounts_update.stake_info*)
+* Slash locked balance for malicious behavior (in order of *validator_accounts_update.slashing_info*)
+* If treasury account was not one of validators and this is first block of an epoch, dispense treasury account reward
+* If this is first block of new version or first block with chunk of new version, apply corresponding migrations
+* If this block do not have chunk for this shard, end process early
+* Process [transactions](RuntimeSpec/Transactions.md) (in order of *transactions*)
+* Process local [receipts](RuntimeSpec/Receipts.md) (in order of *transactions* that generated them)
+* Process delayed [receipts](RuntimeSpec/Receipts.md) (ordered first by block where they were generated, then first local receipts based on order of generating *transactions*,
+then incomming receipts, ordered by *incoming_receipts* order)
+* Process incomming [receipts](RuntimeSpec/Receipts.md) (ordered by *incoming_receipts*)
```
Comment (EgorKulikov) [inline specs/RuntimeSpec/ApplyingChunk.md:29]:
```suggestion
* Process [transactions](Transactions.md) (in order of *transactions*)
* Process local [receipts](Receipts.md) (in order of *transactions* that generated them)
* Process delayed [receipts](Receipts.md) (ordered first by block where they were generated, then first local receipts based on order of generating *transactions*,
then incomming receipts, ordered by *incoming_receipts* order)
* Process incomming [receipts](Receipts.md) (ordered by *incoming_receipts*)
```

File: specs/RuntimeSpec/ApplyingChunk.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Processing chunk
```
Comment (bowenwang1996) [inline specs/RuntimeSpec/ApplyingChunk.md:1]: "Processing chunk" sounds like sometime happens in chain, not runtime. Maybe renaming it to "Apply chunks" would be better?

File: specs/RuntimeSpec/ApplyingChunk.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+* Process [transactions](Transactions.md) (in order of *transactions*)
+* Process local [receipts](Receipts.md) (in order of *transactions* that generated them)
+* Process delayed [receipts](Receipts.md) (ordered first by block where they were generated, then first local receipts based on order of generating *transactions*,
+then incomming receipts, ordered by *incoming_receipts* order)
+* Process incomming [receipts](Receipts.md) (ordered by *incoming_receipts*)
+
+When processing receipts we track gas used (including gas used on migrations). If we use up gas limit, we immidiately stop to process delayed receipts, and for local and incomming
+receipts we add them to delayed receipts)
+
+* If any of the delayed receipts were processed or any new receipts were delayed, update indices of first and last unprocessed receipts in state
```
Comment (bowenwang1996) [inline specs/RuntimeSpec/ApplyingChunk.md:34]: I think more information on how exactly delayed receipts are stored could be helpful (if not already covered in the section on receipts)

File: specs/RuntimeSpec/ApplyingChunk.md (hunk 2 above)
Comment (EgorKulikov) [reply specs/RuntimeSpec/ApplyingChunk.md:1]:
```suggestion
# Applying chunk
```

File: specs/RuntimeSpec/ApplyingChunk.md (hunk 3 above)
Comment (EgorKulikov) [reply specs/RuntimeSpec/ApplyingChunk.md:34]: Done

**Review by bowenwang1996 — APPROVED** [review]

File: specs/RuntimeSpec/ApplyingChunk.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+* If treasury account was not one of validators and this is first block of an epoch, dispense treasury account reward
+* If this is first block of new version or first block with chunk of new version, apply corresponding migrations
+* If this block do not have chunk for this shard, end process early
+* Process [transactions](Transactions.md) (in order of *transactions*)
+* Process local [receipts](Receipts.md) (in order of *transactions* that generated them)
+* Process delayed [receipts](Receipts.md) (ordered first by block where they were generated, then first local receipts based on order of generating *transactions*,
+then incomming receipts, ordered by *incoming_receipts* order)
+* Process incomming [receipts](Receipts.md) (ordered by *incoming_receipts*)
+
+When processing receipts we track gas used (including gas used on migrations). If we use up gas limit, we immidiately stop to process delayed receipts, and for local and incomming
```
Comment (Longarithm) [inline specs/RuntimeSpec/ApplyingChunk.md:31]:
```suggestion
then incoming receipts, ordered by *incoming_receipts* order)
* Process incoming [receipts](Receipts.md) (ordered by *incoming_receipts*)

When processing receipts we track gas used (including gas used on migrations). If we use up gas limit, we immediately stop to process delayed receipts, and for local and incoming
```

File: specs/RuntimeSpec/ApplyingChunk.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+* *validator_proposals*
+* *outgoing_receipts*
+* (executed receipt) *outcomes*
+* *proof*
+
+## Processing order
+
+* If this is first block of an epoch, dispense epoch rewards to validators (in order of *validator_accounts_update.stake_info*)
+* Slash locked balance for malicious behavior (in order of *validator_accounts_update.slashing_info*)
+* If treasury account was not one of validators and this is first block of an epoch, dispense treasury account reward
```
Comment (Longarithm) [inline specs/RuntimeSpec/ApplyingChunk.md:22]:
```suggestion
* If [treasury account](../Economics/README.md#protocol-treasury) was not one of validators and this is first block of an epoch, dispense treasury account reward
```

File: specs/RuntimeSpec/ApplyingChunk.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+and produce following outputs:
+* new state root
+* *validator_proposals*
+* *outgoing_receipts*
+* (executed receipt) *outcomes*
+* *proof*
+
+## Processing order
+
+* If this is first block of an epoch, dispense epoch rewards to validators (in order of *validator_accounts_update.stake_info*)
```
Comment (Longarithm) [inline specs/RuntimeSpec/ApplyingChunk.md:20]:
```suggestion
* If this is first block of an epoch, dispense [epoch rewards](../Economics/README.md#validator-rewards-calculation) to validators (in order of *validator_accounts_update.stake_info*)
```

File: specs/RuntimeSpec/ApplyingChunk.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+## Inputs and outputs
+
+Runtime.apply take following inputs:
+* trie and current state root
+* *validator_accounts_update*
+* *incoming_receipts*
+* *transactions*
+
+and produce following outputs:
```
Comment (Longarithm) [inline specs/RuntimeSpec/ApplyingChunk.md:11]:
```suggestion
Runtime.apply takes following inputs:
* trie and current state root
* *validator_accounts_update*
* *incoming_receipts*
* *transactions*

and produces following outputs:
```

File: specs/RuntimeSpec/ApplyingChunk.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+receipts we add them to delayed receipts)
+
+* If any of the delayed receipts were processed or any new receipts were delayed, update indices of first and last unprocessed receipts in state
+* Remove duplicate validator proposals (for each account we only keep last one in receipts processing order)
+
+Please note that local receipts are receipts generated by transaction where receiver is the same as signer of the transaction
+
+## Delayed receipts
+
+In each block we have maximal amount of Gas we can use to process receipts and migrations, currently 500 TGas. If any incoming or local receipts are not processed due to lack of remaining Gas, they are stored in state with key <pre>TrieKey::DelayedReceipt { id }</pre> where id is unique index for each shard, assigned consecutively Currently sender do not need to stake any tokens to store delayed receipts. State also contain special key <pre>TrieKey::DelayedReceiptIndices</pre> where first and last ids of delayed receipts not yet processed are stored
```
Comment (Longarithm) [inline specs/RuntimeSpec/ApplyingChunk.md:41]:
```suggestion
In each block we have maximal amount of Gas we can use to process receipts and migrations, currently 500 TGas. If any incoming or local receipts are not processed due to lack of remaining Gas, they are stored in state with key <pre>TrieKey::DelayedReceipt { id }</pre> where id is unique index for each shard, assigned consecutively. Currently sender does not need to stake any tokens to store delayed receipts. State also contains special key <pre>TrieKey::DelayedReceiptIndices</pre> where first and last ids of delayed receipts not yet processed are stored.
```

**Review by Longarithm — APPROVED** [review]
LGTM, thank you! Added some minor typo fixes and links which may improve navigation between definitions.

**Pull request merged**